File: src/types/specialAbility/sub/TradeSecret.ts
/**
 * @main TradeSecret
 */

import { validateSchemaCreator } from "../../../validation/schema.js"
import { Errata } from "../../source/_Erratum.js"
import { PublicationRefs } from "../../source/_PublicationRef.js"
import { LocaleMap } from "../../_LocaleMap.js"
import { NonEmptyMarkdown, NonEmptyString } from "../../_NonEmptyString.js"
import { GeneralPrerequisites } from "../../_Prerequisite.js"

/**
 * @title Trade Secret
 */
export type TradeSecret = {
  /**
   * The trade secret's identifier. An unique, increasing integer.
   * @integer
   * @minimum 1
   */
  id: number

  /**
   * The adventure points value of the trade secret.
   * @integer
   * @minimum 1
   */
  ap_value: number

  /**
   * Is this trade secret considered secret knowledge?
   */
  is_secret_knowledge: boolean

  prerequisites?: GeneralPrerequisites

  src: PublicationRefs

  /**
   * All translations for the entry, identified by IETF language tag (BCP47).
   */
  translations: LocaleMap<TradeSecretTranslation>
}

export type TradeSecretTranslation = {
  /**
   * The name of the trade secret.
   */
  name: NonEmptyString

  /**
   * The description of the trade secret.
   */
  description?: NonEmptyMarkdown

  errata?: Errata
}

export const validateSchema = validateSchemaCreator<TradeSecret>(import.meta.url)

File: Sources/OptolithDatabaseSchema/GeneratedTypes/specialAbility/sub/TradeSecret.swift
//
//  TradeSecret.swift
//  OptolithDatabaseSchema
//

public struct TradeSecret: LocalizableEntity {
    /// The trade secret's identifier. An unique, increasing integer.
    public let id: Int

    /// The adventure points value of the trade secret.
    public let apValue: Int

    /// Is this trade secret considered secret knowledge?
    public let isSecretKnowledge: Bool

    public let prerequisites: GeneralPrerequisites?

    public let src: PublicationRefs

    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<TradeSecretTranslation>

    public init(id: Int, apValue: Int, isSecretKnowledge: Bool, prerequisites: GeneralPrerequisites? = nil, src: PublicationRefs, translations: LocaleMap<TradeSecretTranslation>) {
        self.id = id
        self.apValue = apValue
        self.isSecretKnowledge = isSecretKnowledge
        self.prerequisites = prerequisites
        self.src = src
        self.translations = translations
    }

    private enum CodingKeys: String, CodingKey {
        case id = "id"
        case apValue = "ap_value"
        case isSecretKnowledge = "is_secret_knowledge"
        case prerequisites = "prerequisites"
        case src = "src"
        case translations = "translations"
    }
}

public struct TradeSecretTranslation: EntitySubtype {
    /// The name of the trade secret.
    public let name: NonEmptyString

    /// The description of the trade secret.
    public let description: NonEmptyMarkdown?

    public let errata: Errata?

    public init(name: NonEmptyString, description: NonEmptyMarkdown? = nil, errata: Errata? = nil) {
        self.name = name
        self.description = description
        self.errata = errata
    }
}
